File: src/solutions/2024/01/lua/main.lua
local run = require('src.langs.lua.run')
local str = require('src.langs.lua.utils.str')

local options = {
   has_alternate = false,
   has_io = false,
}


local function solve(input, p2)
   local l = {}
   local r = {}
   local hashmap = {}
   for line in input:gmatch('[^\n]+') do
      local pair = str.split(line, '   ')
      l[#l + 1] = pair[1]
      r[#r + 1] = pair[2]
      if (hashmap[pair[1]] == nil) then hashmap[pair[1]] = 0 end
      if (hashmap[pair[2]] == nil) then hashmap[pair[2]] = 0 end
      hashmap[pair[2]] = hashmap[pair[2]] + 1
   end


   local sum = 0
   if p2 then
      for i = 1, #l do
         sum = sum + l[i] * hashmap[l[i]]
      end
   else
      table.sort(l)
      table.sort(r)
      for i = 1, #l do
         sum = sum + math.abs(l[i] - r[i])
      end
   end

   return tostring(sum)
end


---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   return solve(input, false)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   return solve(input, true)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: src/solutions/2024/01/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"os"
	"sort"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func solve(input string, p2 bool) string {

	l := []int64{}
	r := []int64{}
	hashmap := map[int64]int64{}
	for _, line := range strings.Split(input, "\n") {
		pair := strings.Split(line, "   ")
		k, _ := strconv.ParseInt(pair[0], 10, 64)
		v, _ := strconv.ParseInt(pair[1], 10, 64)
		l = append(l, k)
		r = append(r, v)
		if _, ok := hashmap[k]; !ok {
			hashmap[k] = 0
		}
		if _, ok := hashmap[v]; !ok {
			hashmap[v] = 0
		}
		hashmap[v]++
	}

	sum := int64(0)
	if p2 {
		for i := 0; i < len(l); i++ {
			sum += l[i] * hashmap[l[i]]
		}
	} else {
		sort.Slice(l, func(i, j int) bool {
			return l[i] < l[j]
		})
		sort.Slice(r, func(i, j int) bool {
			return r[i] < r[j]
		})
		for i := 0; i < len(l); i++ {
			val := l[i] - r[i]
			if val < 0 {
				sum += -val
			} else {
				sum += val
			}
		}
	}

	return strconv.FormatInt(sum, 10)
}

func part1(input string, isTest bool) string {
	return solve(input, false)
}

func part2(input string, isTest bool) string {
	return solve(input, true)
}
